datafusion/spark/src/function/map/map_function.rs

```diff
@@ -0,0 +1,161 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::array::{ArrayRef, MapArray, StructArray};
+use arrow::buffer::OffsetBuffer;
+use arrow::compute::interleave;
+use arrow::datatypes::{DataType, Field, Fields};
+use datafusion_common::{exec_err, Result};
+use datafusion_expr::{
+    ColumnarValue, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
+};
+
+trait KeyValue<T>
+where
+    T: 'static,
+{
+    fn keys(&self) -> impl Iterator<Item = &T>;
+    fn values(&self) -> impl Iterator<Item = &T>;
+}
+
+impl<T> KeyValue<T> for &[T]
+where
+    T: 'static,
+{
+    fn keys(&self) -> impl Iterator<Item = &T> {
+        self.iter().step_by(2)
+    }
+
+    fn values(&self) -> impl Iterator<Item = &T> {
+        self.iter().skip(1).step_by(2)
+    }
+}
+
+fn to_map_array(args: &[ArrayRef]) -> Result<ArrayRef> {
+    if !args.len().is_multiple_of(2) {
+        return exec_err!("map requires an even number of arguments");
+    }
+    let num_entries = args.len() / 2;
+    let num_rows = args.first().map(|a| a.len()).unwrap_or(0);
+    if args.iter().any(|a| a.len() != num_rows) {
+        return exec_err!("map requires all arrays to have the same length");
+    }
```

Comment:
We shouldn't check this here; we can assume we have valid input and that all arrays are of same length